### PR TITLE
[dev-client][android] refactor

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

I tried to fix integration between `dev-client` and `dev-menu` today (https://www.notion.so/expo/cc8d1a883e59457494cc27ba1dc80ee8?v=9c6126b9e2204e578274cefb9d9c0ba8&p=43ccb9e32f7a4065ac9ec52b93dbf4de). Unfortunately, I have some bad news. It’s not that easy to fix it. I believe, to make it works, the `dev-client` needs to depend on the `dev-menu`. I can’t detect when the launcher changes the current bridge.

Moreover, I’ve also faced some weird issues when I tried to go back to the launcher using the dev-menu button. Sometimes, the app was stuck on the white screen. After a long fight with the RN implementation, I managed to fix that issue. See https://github.com/expo/expo/pull/10824. However, I think that is a good signal to think about using the second Activity to handle the loaded app. It will make our implementation easier and more maintainable. For example, we don’t have to use reflection to change the `ReactDelegate` object that is tied up with the Activity instance. Moreover, I’m still not sure if we cleaned up everything correctly. Also, the refactor should fix all the problems with dev-menu integration.

# How

Use two separate activities. 

# Test Plan

- bare-expo ✅

### CODE REVIEW

File: apps/bare-expo/android/app/src/main/java/dev/expo/payments/MainApplication.java [hunk 1]
```diff
@@ -19,7 +19,7 @@
 import dev.expo.payments.generated.BasePackageList;
 
 public class MainApplication extends Application implements ReactApplication {
-  static final boolean USE_DEV_CLIENT = false;
+  static final boolean USE_DEV_CLIENT = true;
```
Comment (lukmccall) [inline apps/bare-expo/android/app/src/main/java/dev/expo/payments/MainApplication.java:22]:
```suggestion
  static final boolean USE_DEV_CLIENT = false;
```

File: apps/bare-expo/android/app/src/main/java/dev/expo/payments/MainActivity.java [hunk 2]
```diff
@@ -20,15 +20,11 @@ protected String getMainComponentName() {
 
   @Override
   protected ReactActivityDelegate createReactActivityDelegate() {
-    return new ReactActivityDelegate(this, getMainComponentName()) {
+    return DevelopmentClientController.getInstance().wrapReactActivityDelegate(this, new ReactActivityDelegate(this, getMainComponentName()) {
```
Comment (nikki93) [inline apps/bare-expo/android/app/src/main/java/dev/expo/payments/MainActivity.java:23]: Should this be conditioned on `USE_DEV_CLIENT` just in the 'bare-expo' app case?

**Review by nikki93 — COMMENTED** [review]
Nice! Will test this locally and do a deeper review, but had one comment on something I spotted.

**Review by nikki93 — CHANGES_REQUESTED** [review]
I checked out the branch and tried to run 'bare-expo', and everything works fine! When I went back to launcher and re-scanned the QR code for 'bare-expo', it still works, but it actually loaded so quickly that it made me suspect that it was just focusing the already-existent activity. When I went back to launcher again and loaded a QR code from a different app, it still just went back to the activity and showed 'bare-expo'. So it seems like it's not properly reloading / recreating the Activity for each load? Even if it's the same URI we should maybe reload it so we fetch the latest JS (since the developer may be changing the JS while developing their app and are trying to test it).

**Comment by lukmccall** [comment]
You're right 👍  I fixed that :D

**Review by nikki93 — APPROVED** [review]
Just tried again and 'bare-expo' -> back to launcher -> different app works! Code changes look good to me.

**Pull request merged**